File: leet_0735_asteroid_collision.cc
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& a /*asteroids*/)
    {
        vector<int> D;
        int i = 0;
        while (i < (int) a.size())
        {
            // ( -, + ) <queue[-1], incoming>
            // ---> No collision bc they part ways ( << , >> )
            if (D.empty() || a[i] > -1 || D.back() < 0)
            {
                D.push_back(a[i]);
                ++i;
            }
            // ( + , - )<queue[-1], incoming>
            // ---> Is the only possoble colliding scenario
            else if (D.back() + a[i] == 0)
            {
                // either they cancel each other
                D.pop_back();
                ++i;
            }
            else if (D.back() + a[i] < 0)
            {
                // or queue[-1] gets eaten
                D.pop_back();
            }
            else
            {
                ++i;
            }
        }
        return D;
    }
};

File: leet_0735_asteroid_collision.go
func asteroidCollision(/*asteroids*/ A []int) []int {
    var dq []int
    i := 0
	for i < len(A) {
        a := A[i]
		if len(dq) == 0 || a > 0 || dq[len(dq) - 1] < 0 {
            dq = append(dq, a)
            i += 1
        } else if dq[len(dq) - 1] + a == 0 {
            dq = dq[:len(dq) - 1]
            i += 1
        } else if dq[len(dq) - 1] + a < 0 {
            dq = dq[:len(dq) - 1]
        } else {
            i += 1
        }
	}
	return dq
}
